[cd18] Scenario#1 - 프로젝트 설정 및 공연 정보 리스트/상세 API 구현

CodeDiary18 opened the pull request:
## Implemented tasks
### 완료한 작업 요약
- 멀티 모듈 레이어 설계
- 공연 정보 리스트/상세 API 구현

<br>

### 프로젝트 모듈 구조
> common 모듈은 설명에서 제외
- 기본 컨셉 : 헥사고날 아키텍처 + 멀티 모듈 방식

- `web-api` 모듈
    - 클라이언트로부터의 HTTP 요청 처리
    - API 엔드포인트 정의
    - 요청 데이터의 유효성 검사
    - `domain` 모듈의 서비스 계층을 호출하여 비즈니스 로직 실행
- `domain` 모듈
    - 애플리케이션의 핵심 비즈니스 로직 담당
    - 하위에 domain 패키지와 application 패키지 존재  
       (application 모듈을 분리할까 고민했는데 우선은 domain 모듈 안에서 패키지로 구분하기로 함)
    - domain 패키지
        - 애플리케이션의 핵심 도메인 모델과 비즈니스 규칙 정의
        - entity 정의, service와 repository interface 정의
    - application 패키지
        - 도메인 패키지에서 정의된 비즈니스 로직을 조합하여 애플리케이션의 특정 기능 구현
        - 서비스 구현, 트랜잭션 관리, 외부 시스템 통합
- `infra` 모듈
    - `domain` 모듈에서 정의한 repository 인터페이스를 구현 or 외부 서비스와의 통신 담당
    - **persistence**: repository 인터페이스 구현 (DB 통신)
    - **cache**: 캐시 시스템 통합 (Redis 등)
    - **messaging**: 메시지 큐 통합 (Kafka, RabbitMQ 등)
    - **storage**: 파일 저장소 통합 (AWS S3 등)

**동작 흐름**

1. 클라이언트 → `web-api` (요청 처리)
2. `web-api` → `domain` (비즈니스 로직 호출)
3. `domain` → `infra` (데이터 접근 또는 외부 시스템 통합)
4. `infra` → `domain` → `web-api` → 클라이언트 (응답 반환)

<br><br>

## Request to reviewer
- 전체 프로젝트 구조 위주로 리뷰 부탁드립니다
  - 모듈간(?) 의존성 주입 최적화 부분에 대해서도 팁이 있다면 공유 부탁드립니다😊
- 테스트 코드는 몇가지 방법을 시도중이라 정리되면 커밋할 예정입니다

<br>

#### Reference
- [이번 시나리오 관련 생각 정리 노션 링크](https://daily-wednesday-40d.notion.site/Scenario-1-16a712658b25803797c6d391652168b0)

CodeDiary18 committed:
chore: setting MySQL

CodeDiary18 committed:
chore: table ddl script

CodeDiary18 committed:
feat: define entity for performance info

CodeDiary18 committed:
chore: 공연정보 목록 & 상세 API 설계

CodeDiary18 committed:
fix: redefine entity

CodeDiary18 committed:
chore: querydsl setting

CodeDiary18 committed:
chore: downgrade springboot & swagger version

to resolve the db connection issue

CodeDiary18 committed:
refactor: update api response dto detail

CodeDiary18 committed:
feat: define PageParam and PageRequest

CodeDiary18 committed:
feat: define PerformanceInfo Service & Repository interface

CodeDiary18 committed:
chore: setting for db connection

- add dependency
- config EntityScan range

CodeDiary18 committed:
chore: add db dummy script

CodeDiary18 committed:
feat: implement PerformanceInfo Service & Repository

for performance info list and detail api

CodeDiary18 committed:
feat: inject PerformanceInfoService into PerformanceInfoController

CodeDiary18 committed:
feat: implement PerformanceInfo list/detail logic

CodeDiary18 committed:
refactor: define paging extension function for querydsl

CodeDiary18 committed:
refactor: setting Entity id initial value

Octoping925 commented on infra/src/main/kotlin/com/cd18/infra/persistence/config/QueryDslExtensions.kt:
```diff
@@ -0,0 +1,11 @@
+import com.querydsl.jpa.impl.JPAQuery
+import org.springframework.data.domain.Pageable
+
+fun <T> JPAQuery<T>.applyPagination(pageParam: PageParam): JPAQuery<T> {
+    val pageable: Pageable = Pageable.ofSize(pageParam.size).withPage(pageParam.page)
+    return this.offset(pageable.offset)
+        .limit(pageable.pageSize.toLong())
+}
```
top level function 활용 좋네 :+1:

Octoping925 commented on infra/src/main/kotlin/com/cd18/infra/persistence/model/PerformanceDate.kt:
```diff
@@ -0,0 +1,22 @@
+    @Column(name = "perf_id")
+    var performanceInfoId: Long,
+    @Column(name = "perf_time")
+    var startTime: Date,
```
뭔가 LocalDateTime을 쓰는건 많이 봤는데, `java.sql.Date`는 새로운 것 같네
어떤 이유로 사용한건지 들어보고 싶어

CodeDiary18 replied:
실수였습니다.. 허허😅

Octoping925 commented on web-api/build.gradle.kts:
```diff
@@ -6,13 +6,23 @@ plugins {
+    runtimeOnly("com.mysql:mysql-connector-j:8.4.0")
+    implementation("com.zaxxer:HikariCP")
+    implementation("org.hibernate.validator:hibernate-validator:8.0.1.Final")
+    implementation("org.glassfish:jakarta.el:4.0.2")
```
DB와 연관된 의존성은 infra 쪽에 가는게 맞지 않을까?
web-api는 Web과 관련된 의존성만 있으면 좋을 것 같은데

Octoping925 commented on domain/build.gradle.kts:
```diff
@@ -0,0 +1,24 @@
+
+dependencies {
+    implementation(project(":common"))
+    implementation("org.springframework.boot:spring-boot-starter-web")
```
도메인이 스프링을 아는 건 `@Service` 이런 것들 때문이지?
다른 의존성은 빼고 컴포넌트 관련 의존성만 얻는 방법이 있어

`compileOnly("org.springframework:spring-context")`

CodeDiary18 replied:
참고해서 수정 완료했습니다~ 땡큐🤗

Octoping925 commented on domain/build.gradle.kts:
```diff
@@ -0,0 +1,24 @@
+dependencies {
+    implementation(project(":common"))
+    implementation("org.springframework.boot:spring-boot-starter-web")
+    implementation("org.jetbrains.kotlin:kotlin-reflect")
+    testImplementation("org.springframework.boot:spring-boot-starter-test")
+    testImplementation("org.jetbrains.kotlin:kotlin-test-junit5")
+    testRuntimeOnly("org.junit.platform:junit-platform-launcher")
```
전체 모듈에 공통적으로 필요한 친구들은 최상단 build.gradle로 빼도 괜찮을듯?

Octoping925 reviewed (approved):
전반적으로 깔끔하게 잘 짠 것 같아!
멀티모듈 의존성 관련해서 조금 최적화할 부분이 보이는 것 같기는 하네

CodeDiary18 committed:
chore: move common dependencies to the root build.gradle

CodeDiary18 committed:
refactor: change startTime from Date to LocalDateTime in PerformanceDate

CodeDiary18 committed:
chore: remove spring-boot-starter-web dependency in infra module

CodeDiary18 committed:
chore: move db dependencies to infra module

CodeDiary18 committed:
chore: refactor spring dependency in domain module

- add spring-web dependency for HttpStatus
- add spring-context for Spring container features
